---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Users/mapostolides/Drive/fusion-pipeline/genometools_introns')
```

```{r, message=FALSE, echo=FALSE}
# check to ensure all needed packages are installed
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
# if (!requireNamespace("GEOmetadb", quietly = TRUE))
#   BiocManager::install("GEOmetadb")
# if (!requireNamespace("edgeR", quietly = TRUE))
#   BiocManager::install("edgeR")
if (!requireNamespace("biomaRt", quietly = TRUE))
  BiocManager::install("biomaRt")
# if (!requireNamespace("ComplexHeatmap", quietly = TRUE))
#   BiocManager::install("ComplexHeatmap")
# if (!requireNamespace("gProfileR", quietly = TRUE))
#   BiocManager::install("gProfileR")
# if (!requireNamespace("gprofiler2", quietly = TRUE))
#   BiocManager::install("gprofiler2")
if (!requireNamespace("stringr", quietly = TRUE))
  install.packages("stringr")
if (!requireNamespace("kableExtra", quietly = TRUE))
  install.packages("kableExtra")
if (!requireNamespace("tidyr", quietly = TRUE))
  install.packages("tidyr")
```


```{r, message=FALSE, echo=FALSE}
# load needed libraries
# library("GEOmetadb")
# library("edgeR")
library("biomaRt")
# library("tidyr")
library("dplyr")
# library("ComplexHeatmap")
# library(circlize)
# library("gprofiler2")
library("stringr")
library(kableExtra)
```

```{r}
#load vega annot file, for adding in gene names missing from biomart
vega.annot <- paste0( getwd(), "/","feb2014.vega.mart_export.txt")
vega.annot <- read.csv(file=vega.annot, header = T, stringsAsFactors = F, sep="\t", na.strings = "")

#vega.annot <- vega.annot[vega.annot$Corresponding.Ensembl.Gene.ID.s. == "ENSG00000236601",] 
vega.annot <- unique(vega.annot[c("External.Gene.ID", "Corresponding.Ensembl.Gene.ID.s.")])
#remove all rows which don't have an ENSG ID
vega.annot <- vega.annot[grepl("ENSG[0-9]+", vega.annot$Corresponding.Ensembl.Gene.ID.s.),]
#rename cols
colnames(vega.annot)[colnames(vega.annot) == "Corresponding.Ensembl.Gene.ID.s."] <- "ensembl_gene_id"
colnames(vega.annot)[colnames(vega.annot) == "External.Gene.ID"] <- "gene_name"
vega.annot

```


```{r map_ensg_to_hgnc, echo=FALSE, message=FALSE, warning=FALSE}
#LOAD BED FILES I GENERATED
cols <- c("chr", "start", "end", "idx", "TMP", "strand", "V6", "type", "V7", "transcript_id" )
# Read in .bed file generated from "Homo_sapiens.GRCh37.87.chr.INTRONS.gff3"
#STPG1.bed <- paste0( getwd(), "/", "STPG1.ENST00000003583.sorted.bed")
#STPG1.bed <- read.csv(file=STPG1.bed, header = FALSE, stringsAsFactors = F, sep="\t", na.strings = "")

total.bed <-  paste0( getwd(), "/", "Homo_sapiens.GRCh37.87.chr.INTRONS.sorted.bed")
total.bed <- read.csv(file=total.bed, header = FALSE, stringsAsFactors = F, sep="\t", na.strings = "")
colnames(total.bed) <- cols

```


```{r}
#LOAD ensemblToGeneName.Ensembl_Genes.Sept-22-2020.txt
ensemblToGeneName <- paste0( getwd(), "/", "ensemblToGeneName.Ensembl_Genes.Sept-22-2020.txt")
ensemblToGeneName <- read.csv(file=ensemblToGeneName, header = TRUE, stringsAsFactors = F, sep="\t", na.strings = "")
names(ensemblToGeneName)[names(ensemblToGeneName) == 'X.name'] <- 'transcript_id'
names(ensemblToGeneName)[names(ensemblToGeneName) == 'value'] <- 'gene_name'
```



```{r map_ensg_to_hgnc, echo=FALSE, message=FALSE, warning=FALSE}
# SET BED
bed <- total.bed
#1	24683494	24685032	.	.	-	ensembl_havana	three_prime_UTR . Parent=transcript:ENST00000003583

# Remove all rows which do not have an ENST ID
bed <- bed[grepl("ENST[0-9]+", bed$transcript_id),]

# THIS BLOCK just confirms that all rows removed due to not having ENST are not causing us to lose any ENSGs
#Store all rows which have no ENST IDs
total.bed.no_ENSTs <- anti_join(total.bed, bed)
# Get all rows from "total.bed.no_ENSTs" which have ENSG IDs
total.bed.no_ENSTs.with_ENSGs <- total.bed.no_ENSTs[grepl("ENSG[0-9]+", total.bed.no_ENSTs$transcript_id),]
#remove all characters from column 10 except the ENSG IDs 
total.bed.no_ENSTs.with_ENSGs$gene_id <- regmatches(total.bed.no_ENSTs.with_ENSGs$transcript_id, regexpr("ENSG[0-9]+",total.bed.no_ENSTs.with_ENSGs$transcript_id) )
# all(total.bed.no_ENSTs.with_ENSGs$gene_id %in% bed.ensemblToGeneName$gene_id) --> TRUE

#remove all characters from column 10 except the ENST IDs 
bed$transcript_id <- regmatches(bed$transcript_id, regexpr("ENST[0-9]+",bed$transcript_id) )

# Retrieve conversion using biomart or conversion stash 
conversion_stash <- "conversion_stash.full_bed.GRCh37.rds"
if(file.exists(conversion_stash)){
  conversion <- readRDS(conversion_stash)
} else {
    # Biomart query
    ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", GRCh=37)
    #MAP ENST TO ENSG SYMBOLS
    conversion <- getBM(attributes = c("ensembl_transcript_id", "ensembl_gene_id","hgnc_symbol"), 
                         filters = c("ensembl_transcript_id"), values = unique(bed$transcript_id), mart = ensembl)
    saveRDS(conversion, conversion_stash)
}

# Map conversion data to bed file, merge using join
names(conversion)[names(conversion) == 'ensembl_transcript_id'] <- 'transcript_id'
names(conversion)[names(conversion) == 'ensembl_gene_id'] <- 'gene_id'
#names(conversion)[names(conversion) == 'hgnc_symbol'] <- 'gene_name'

# Map ENSTs to ENSGs where possible, removing entries in "bed" which lack ENSGs
joined.bed <- inner_join(bed, conversion)
bed.ensemblToGeneName <-  left_join(joined.bed, ensemblToGeneName, by="transcript_id")

saveRDS(bed.ensemblToGeneName, file="/Users/mapostolides/Drive/fusion-pipeline/genometools_introns/bed.ensemblToGeneName.new_bed.Rds")
#
#feature_types <- unique(bed.ensemblToGeneName$type)

# Filter feature type to be only among c("exon", "intron", "CDS", "three_prime_UTR","five_prime_UTR", "snoRNA", "snRNA", "lincRNA", "miRNA", "rRNA"  )
#features <- c("exon", "intron", "CDS", "three_prime_UTR","five_prime_UTR", "snoRNA", "snRNA", "lincRNA", "miRNA", "rRNA"  )
features <- c("exon", "intron", "CDS", "three_prime_UTR","five_prime_UTR")
bed.ensemblToGeneName.feature_subset <- bed.ensemblToGeneName[bed.ensemblToGeneName$type %in% features,]

# bed.ensemblToGeneName[bed.ensemblToGeneName$gene_id == "ENSG00000009790" & bed.ensemblToGeneName$transcript_id == "ENST00000487271" & bed.ensemblToGeneName$type == "five_prime_UTR",]

features<- c("processed_transcript", "pseudogenic_transcript", "pseudogene","lincRNA", "miRNA","mRNA", "processed_pseudogene", "transcript","snRNA","aberrant_processed_transcript", "NMD_transcript_variant",  "snoRNA","rRNA", "nc_primary_transcript", "V_gene_segment",  "J_gene_segment", "C_gene_segment")
bed.ensemblToGeneName.RNA_types <- bed.ensemblToGeneName[bed.ensemblToGeneName$type %in% features,]

# bed.ensemblToGeneName.snRNAs <- bed.ensemblToGeneName[bed.ensemblToGeneName$transcript_id %in% unique(bed.ensemblToGeneName[bed.ensemblToGeneName$type == "snRNA",]$transcript_id),]
# bed.ensemblToGeneName.snoRNAs <- bed.ensemblToGeneName[bed.ensemblToGeneName$transcript_id %in% unique(bed.ensemblToGeneName[bed.ensemblToGeneName$type == "snoRNA",]$transcript_id),]
# bed.ensemblToGeneName.rRNAs <- bed.ensemblToGeneName[bed.ensemblToGeneName$transcript_id %in% unique(bed.ensemblToGeneName[bed.ensemblToGeneName$type == "rRNA",]$transcript_id),]
# bed.ensemblToGeneName.J_gene_segment <- bed.ensemblToGeneName[bed.ensemblToGeneName$transcript_id %in% unique(bed.ensemblToGeneName[bed.ensemblToGeneName$type == "J_gene_segment",]$transcript_id),]
```

```{r}
# Index exons, introns, utr5, utr3 

# Process and index each transcript_id separately
transcript_ids <- unique(bed.ensemblToGeneName.feature_subset$transcript_id)
#CONFIRMING that # of transcript_ids doesn't change
#length(transcript_ids) --> 196354
#length(unique(bed$transcript_id)) --> 196354

#transcript_ids <- c("ENST00000591849", "ENST00000591534", "ENST00000586385")
#transcript_ids <- c("ENST00000431554")
#transcript_ids <- c("ENST00000593055")

# id <- "ENST00000431554"
# id <- "ENST00000593055" # ACACA
# transcript <- bed.ensemblToGeneName[bed.ensemblToGeneName.feature_subset$transcript_id == id,]
# transcript

#feature_types <- c("exon", "intron", "CDS", "three_prime_UTR","five_prime_UTR" )

file.to_write <- "/Users/mapostolides/Drive/fusion-pipeline/genometools_introns/new_bed.total.bed"
#file.to_write <- "/Users/mapostolides/Drive/fusion-pipeline/genometools_introns/tmp.bed"
#file.to_write <- "/Users/mapostolides/Drive/fusion-pipeline/genometools_introns/BRCA.bed"
#file.to_write <- "/Users/mapostolides/Drive/fusion-pipeline/genometools_introns/ENST00000431554.bed"
if(file.exists(file.to_write) ) {file.remove(file.to_write)}

#START CLOCK
ptm <- proc.time()

# Index each transcript feature, incrementing when an intron is passed
for (id in transcript_ids){
  coding <- 0
  transcript <- bed.ensemblToGeneName.feature_subset[bed.ensemblToGeneName.feature_subset$transcript_id == id,]
  # Remove exons if coding gene, since "exon" and "CDS" are duplicates of one another
  if ("CDS" %in% transcript$type){
    transcript <- transcript[!transcript$type == "exon",]
  }
  # Order features by increasing bp 
  transcript <- transcript[order(transcript$start, decreasing = FALSE),]
  # Index features
  idx <- 0
  for (i in 1:nrow(transcript)){
    transcript[i,]$idx <- idx
    if (transcript[i,]$type == "intron"){
      #print(transcript[i,]$type)
      idx <- idx + 1
    }
  }
  # REFORMAT TRANSCRIPT
  #Change strand info (+ --> f, - --> r)
  if (unique(transcript$strand) == "+"){
    transcript$strand <- 'f'
  } else if  (unique(transcript$strand) == "-"){
    transcript$strand <- 'r'
  } else {
    errorCondition("Strand info for this transcript is inconsistent")
  }
  #Add "chr" prefix to chromosomes
  transcript$chr <- sapply("chr", paste0,  transcript$chr)
  #Change CDS --> cds
  if ("CDS" %in% unique(transcript$type)){transcript[transcript$type == "CDS",]$type <- "cds"}
  # Change "three_prime_UTR" --> "utr3"
  if ("three_prime_UTR" %in% unique(transcript$type)){transcript[transcript$type == "three_prime_UTR",]$type <- "utr3"}
  #Change "five_prime_UTR"  --> "utr5"
  if ("five_prime_UTR" %in% unique(transcript$type)){transcript[transcript$type == "five_prime_UTR",]$type <- "utr5"}
  transcript <- transcript[c("chr", "start", "end", "transcript_id", "type", "idx", "strand", "gene_name", "gene_id" )]
  write.table(transcript, file.to_write, append=TRUE, sep="\t", quote=F,  row.names=F, col.names=F)

}

#STOP CLOCK
time <- proc.time() - ptm
time
```


```{r}



#FIND DIFFERENCE IN GENE NAMES BETWEEN 2 BED FILES
ens.genes <- unique(ens.known.genes$gene_name)
joined.bed.genes <- unique(joined.bed$gene_name)

# vega.genes <- unique(vega.annot$gene_name)
ens.only <- anti_join(ens.known.genes, joined.bed,by="gene_name")
ens.only.genes <- unique(ens.only$gene_name)
joined.only <- anti_join(joined.bed,ens.known.genes,by="gene_name")
joined.only.genes <- unique(joined.only$gene_name)

# length(ens.genes) = 56022
# length(ens.only.genes) = 23600
# length(joined.bed.genes) = 36933
# length(joined.only.genes) = 4511

#Find overlap between Vega genes and ens.only.genes
length(intersect(vega.genes, ens.only.genes))
#Find genes in ens not in vega
ens.only.no_vega <- anti_join(ens.only, vega.annot, joined.bed,by="gene_name")
ens.only.no_vega.genes <- unique(ens.only.no_vega$gene_name)
length(ens.only.no_vega.genes)
```


```{r}
#FIND DIFFERENCE IN ENSG IDs BETWEEN 2 BED FILES

#number of unique ENSG IDs in ens.genes
# length(unique(ens.known.genes$gene_id)) = 60234
#number of unique ENSG IDs in joined.bed
# length(unique(joined.bed$gene_id)) = 52849

ens.known.unique_ENSGs <- anti_join(ens.known.genes, joined.bed, by='gene_id')
# length(unique(ens.known.unique_ENSGs$gene_id)) = 8209

```

